### ISSUE TITLE
Data loss issue in ssl_write_real(), which can be escalated into DoS/authorization bypass 

### ISSUE THREAD

**Issue by pfalcon** [ARMmbed/mbedtls#1356]
Note: This is just a template, so feel free to use/remove the unnecessary things

### Description
- Type: Bug
- Priority: Major (?)

---------------------------------------------------------------
## Bug

**OS**  
linux/any

**mbed TLS build:**  
2.7.0/32605dc83042d737e715a685e53176388d73540e
Configuration: Non-blocking underlying transport is used (MBEDTLS_ERR_SSL_WANT_WRITE returned from user send() callback).  

**Expected behavior**   
No unreported sent data loss occur. 

**Actual behavior**
Possibility for loss of application data requested for sending out, if non-blocking sockets (or other communication means) are used.

**Steps to reproduce**  
This is based on the mbedTLS code review, preparing to integrate mbedTLS into a system (Zephyr RTOS socket-based communication). The below describes the issue.

The culprit is in the function ssl_write_real(), line 7142 of library/ssl_tls.c at the time of writing ("development" branch HEAD is 32605dc83042d737e715a685e53176388d73540e): https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_tls.c#L7142

ssl_write_real() is called from mbedtls_ssl_write() which is public API for mbedTLS.

As can be seen at https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_tls.c#L7187 on the "success" path, it returns `len`, which is the input `len` param, possibly capped by the max_len variable.

This return is returned even if condition https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_tls.c#L7166
`if( ssl->out_left != 0 )` evaluates to true. This branch of the `if/else` statement deals with the case when leftover bytes from the previous TLS record are still in output buffer. This condition can happen if the underlying network stream (e.g. socket) is in non-blocking mode (MBEDTLS_ERR_SSL_WANT_WRITE returned from user-supplied send() callback, which would correspond to POSIX EAGAIN errno code). As the statement is `if/else`, if the statement "mbedtls_ssl_flush_output()" in the `if` branch finishes
successfully (signifying that all leftover bytes were successfully spooled to the network), then `else` branch (which deals with sending the new user data passed in by the user to the current invocation) will not be executed.

But in this case, value of `len` still will be returned! In other words, none of the application data will be sent, but return value will indicate that it (capped to the value of max_len) is sent successfully.

This is effectively a data loss issue. It would affect with non-zero probability high-performance asynchronuous applications using non-blocking sockets. A typical case when EAGAIN is returned is when there're no underlying network buffers to accept more user data. Availability of network buffers varies by entire system's network load. That means that the condition can be caused externally, turning this issue into Denial of Service. For example, an attacker, knowing that there're mbedTLS communication channels running on on a system, can exert network load on the system to cause network buffer exhaustion. This doesn't have to be exhaustive, flood-style traffic, especially for small systems, which are direct usecases for mbedTLS. Sending incomplete, partial data from client to server (or among systems) should lead todenial of service(s) on its own. An example of specific case of denial may include: If authorization data is omitted when performing a request - access will be denied, for no apparent to a user reason.

Worst case, sending partial data may lead to authorization bypass/incorrect authorization. For example, if there're 2 users with identifiers AAAA and AAAABBBB, the second identifier is being sent, which happens to be transmitted by application layer in parts (due to buffering, etc.), then "AAAA" can be sent, while "BBBB" transparently lost, causing incorrect authorization.

**Comment by pfalcon** [comment]
As an additional information, the initial motivation for looking at these functions was trying to figure out how to flush mbedTLS output buffer. My first surprise was finding out that mbedTLS doesn't really "gather-buffers" user data, i.e. doesn't try to collect an output buffer-ful of user data, and send it at once, and instead creates a TLS record for each call to mbedtls_ssl_write() (i.e. even for 1 byte). While exploring that I noticed the issue described here, and, because I looked at 2.6.0 initially, a buffer overflow issue which turned to be fixed for 2.7.0 in https://github.com/ARMmbed/mbedtls/commit/0b7b83fd9 .

Generally, it took ~5 mins to discover these, arguably pretty obvious issues. It was a bit hard to believe in that, so I tried to see what I may be missing (I of course still accept the possibility that the above analysis has error). I traced it to 1fd00bfe82af7c52f3d0d042cbd97e554d84b9ad which "fixed" the issues, and then 887bd502d2f2052816adbb21be254b254ad8d9a0 reverted it back, because it supposed did more harm than good. Both commits were yet at PolarSSL times, and refer to its bugtracker and ticket no.18 there for more info. As far as I can tell, it's no longer accessible (btw, another reason to avoid just referring to just ticket numbers, instead of explaining the details of the issues directly in the commit message). Maybe, there're people around which can shed some light on that.

Thanks.

**Comment by carnil** [comment]
The issue was assigned CVE-2018-1000061

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-2093

**Comment by RonEld** [comment]
Hi @pfalcon Thank you for your information.
I am having some trouble following your use case. According to the [ function's description](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/ssl.h#L2512),:
~~~
* \note           When this function returns MBEDTLS_ERR_SSL_WANT_WRITE/READ,
 *                 it must be called later with the *same* arguments,
 *                 until it returns a positive value.
~~~
But if I understand your description correct, in previous call, the error `MBEDTLS_ERR_SSL_WANT_WRITE ` is returned, but you are trying to send a new application data. Why do you think there is partial sending?

**Comment by pfalcon** [comment]
@RonEld : Well, a hat-eating moment for me. With your simple hint, retracing the processing stages, I see that I was wrong and the code works correctly. The only reason why I thought the processing was incorrect apparently boils down to: I thought that on the initial call, when user data is being copied into output buffer and encoded int a TLS record, the return value is "success" (i.e. the length). But why I thought that way, if the code clearly doesn't do it? The only explanation would be "that's how I'd do it". Well, that way might be more "efficient" (user's buffer is "released", and an application may start to fill with with data sooner), but mbedTLS's behavior is more conservative and leaves less possibility for data loss. Still, I made conclusions out of my wishful thinking.

Arguably, this handling may be confusing, as commits from 2011 quoted above show. So, I'd like to take a chance to try to clarify it by adding code comments.

With all the above, it's hard to explain why I got so bold as to submit a CVE request for the above. I feel stupid and childish. It will be a good lesson for me, that real vulnerability analysis should be done with explicit analysis of any assumptions, should be backed by dynamic analysis using proof of concept code, etc. My apologies to all the parties to which that CVE submission caused inconveniences :-(((.

Summary:

* This issue is **INVALID**.
* I submitted https://github.com/ARMmbed/mbedtls/pull/1370 in the hope to make something good out of it still.
* I'll work towards marking  CVE-2018-1000061 as REJECTED.
